derp/derp_server.go

```diff
@@ -889,6 +981,96 @@ func (c *sclient) sendPacket(srcKey key.Public, contents []byte) (err error) {
 	return err
 }
 
+// AddPacketForwarder registers fwd as a packet forwarder for dst.
+// fwd must be comparable.
+func (s *Server) AddPacketForwarder(dst key.Public, fwd PacketForwarder) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	if s.clientsElsewhere == nil {
+		s.clientsElsewhere = map[key.Public]PacketForwarder{}
+	}
+	if prev, ok := s.clientsElsewhere[dst]; ok {
+		if prev == fwd {
+			// Duplicate registration of same forwarder. Ignore.
+			return
+		}
+		if m, ok := prev.(multiForwarder); ok {
+			if _, ok := m[fwd]; !ok {
+				// Duplicate registration of same forwarder in set; ignore.
+				return
+			}
+			m[fwd] = m.maxVal() + 1
+			return
+		}
+		// Otherwise, the existing value is not a set and not a dup, so make it a set.
+		fwd = multiForwarder{
+			prev: 1,
+			fwd:  2,
+		}
+	}
+	s.clientsElsewhere[dst] = fwd
+}
+
+// RemovePacketForwarder removes fwd as a packet forwarder for dst.
+// fwd must be comparable.
+func (s *Server) RemovePacketForwarder(dst key.Public, fwd PacketForwarder) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	v, ok := s.clientsElsewhere[dst]
+	if !ok {
+		return
+	}
+	if m, ok := v.(multiForwarder); ok {
+		if len(m) < 2 {
+			panic("unexpected")
+		}
+		delete(m, fwd)
+		// If fwd was in m and we no longer need to be a
+		// multiForwarder, replace the entry with the
+		// remaining PacketForwarder.
+		if len(m) == 1 {
+			var remain PacketForwarder
+			for k := range m {
+				remain = k
+			}
+			s.clientsElsewhere[dst] = remain
+		}
+		return
+	}
+	if v != fwd {
+		// Delete of an entry that wasn't in the map.
+		return
+	}
+	delete(s.clientsElsewhere, dst)
+}
+
+// multiForwarder is a set of forwarders. It's used in the rare cases that a client
+// is connected to multiple DERP nodes in a region. The value is unique connection
+// number; the lowest one has been seen the longest. It's used to make sure we
+// forward packets consistently to the same node and don't pick randomly.
+type multiForwarder map[PacketForwarder]uint8
```

Comment:
Oh I see, it's a packet tee. From the name, I assumed it was to handle the case of >2 derp servers in a region, rather than clients incorrectly connected multiple times. Maybe name this duplicatingForwarder, or something that makes it explicit that it's copying packets rather than just forwarding?

Reply:
It's not teeing. It only forwards to 1. If user1 is connected to both 2a and 2b, if user2 sends to user1 on 2c, I don't want to randomly pick whether to forward to 2a and 2b each time.